### PR TITLE
A first crack at process memory <=> git-theta storage

### PR DESCRIPTION
**Pull request by blester125**
Demo:

```python
import os
import torch
import git_theta
from git_theta.api import save, load
import numpy as np

def make_model():
  return torch.nn.Sequential(
      torch.nn.Linear(10, 5),
      torch.nn.ReLU(),
      torch.nn.Linear(5, 2))

model = make_model()
model2 = make_model()

path = "model.pt"
sha = save(model.state_dict(), path, "model from py")

assert not os.path.exists(path)
assert not np.allclose(model[0].weight.detach().numpy(), model2[0].weight.detach().numpy())

m = load(sha, path)
model2.load_state_dict(m)

assert np.allclose(model[0].weight.detach().numpy(), model2[0].weight.detach().numpy())
assert not os.path.exists(path)
```

both asserts pass meaning we were able to save and load the model without it actually being written into the working tree!

First approximation of https://github.com/r-three/git-theta/issues/210

### CODE REVIEW

File: git_theta/api.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import git_theta
+from git_theta import checkpoints, filters, git_utils, metadata, utils
+
+
+# TODO: Should we have a `materialize` flag that writes the model to disk so you
+# don't have git status showing a diff?
+def save(
+    state_dict,
+    path: str,
+    commit_msg: Optional[str] = None,
```
Comment (craffel) [inline git_theta/api.py:18]: Do we really want to make this optional? Git treats it as non-optional.

File: git_theta/api.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    else:
+        # Commit directly from python
+        sha = repo.index.commit(commit_msg)
+    return sha
+
+
+# TODO: Should we add something like a `checkout` parameter that actually checks
+# the model out so we are at the right commit after using this model?
+def load(
+    sha: git.Commit,
```
Comment (craffel) [inline git_theta/api.py:56]: Should we try to be fancy and allow loading a file via a tag rather than a commit hash?

File: git_theta/checkpoints/base.py [hunk 3]
```diff
@@ -49,6 +49,15 @@ def load(cls, checkpoint_path):
             should be numpy arrays.
         """
 
+    @classmethod
+    @abstractmethod
+    def from_framework(cls, model_dict):
+        """Convert a checkpoint from the native framework format to git-thetas."""
+
+    @abstractmethod
+    def to_framwork(self):
```
Comment (craffel) [inline git_theta/checkpoints/base.py:58]: framework

File: git_theta/checkpoints/flax_checkpoint.py [hunk 4]
```diff
@@ -16,6 +16,13 @@ class FlaxCheckpoint(Checkpoint):
     def load(cls, checkpoint_path, **kwargs):
         return serialization.msgpack_restore(checkpoint_path.read())
 
+    @classmethod
+    def from_framwork(cls, model_dict):
```
Comment (craffel) [inline git_theta/checkpoints/flax_checkpoint.py:20]: framework

File: git_theta/checkpoints/flax_checkpoint.py [hunk 5]
```diff
@@ -16,6 +16,13 @@ class FlaxCheckpoint(Checkpoint):
...
+        return cls(model_dict)
+
+    def to_framwork(self):
```
Comment (craffel) [inline git_theta/checkpoints/flax_checkpoint.py:23]: framework

File: tests/end2end/inprocess/test.sh [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+echo "Making model ${MODEL_NAME}"
+INIT_MODEL=`python ../${MODEL_SCRIPT} --action init --seed 1337 --model-name=${MODEL_NAME}`
+
+python test.py
+if [[ "$?" != 0 ]]; then
+    exit 1
+fi
+echo "Verifying that the same model saved in different paths match"
+python ../verify.py --old-model should_match_1.pt --new-model should_match_2.pt
```
Comment (craffel) [inline tests/end2end/inprocess/test.sh:21]: Is there a reason we can't use the functionality from verify.py directly in test.py rather than doing it separately in a shell script? Or is there a benefit to doing it separately?

File: git_theta/filters.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
+"""Clean and Filter functions."""
```
Comment (craffel) [inline git_theta/filters.py:1]: Did these change at all? I'm not inspecting them deeply...

File: tests/end2end/inprocess/test.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+print("committing the same model to different paths")
+model_1_sha = git_theta.save(model, "model_1.pt")
+model_2_sha = git_theta.save(model, "model_2.pt")
+print("committing the changed model to the same path.")
+model_1_1_sha = git_theta.save(updated_model, "model_1.pt")
+
+print("Making sure the models we not saved to disk.")
+assert not os.path.exists("model_1.pt")
```
Comment (craffel) [inline tests/end2end/inprocess/test.py:25]: Re: your comment above: "TODO: Should we have a `materialize` flag that writes the model to disk so you
# don't have git status showing a diff?" It kind of creeps me out that are referring to things via paths that aren't actually ever materialized on disk...

File: git_theta/api.py (hunk 1 above)
Comment (blester125) [reply git_theta/api.py:18]: We have a default value of `Committing ${path}`, I assume if we make it mandatory there would be a lot of `ups`/`update` commits.

I think the advantage of not having a default is that function sig basically matches torch.save, but we can remove the default if we think it will make people add more informative commits.

File: git_theta/api.py (hunk 2 above)
Comment (blester125) [reply git_theta/api.py:56]: GitPython already supports commit lookup from either a sha1 or a tag so we actually already support that. I've updated argument names to reflect that.

File: git_theta/checkpoints/base.py (hunk 3 above)
Comment (blester125) [reply git_theta/checkpoints/base.py:58]: This explains why the abstractmethod part wasn't working right lol.

Fixed

File: git_theta/checkpoints/flax_checkpoint.py (hunk 5 above)
Comment (blester125) [reply git_theta/checkpoints/flax_checkpoint.py:23]: fixed

File: git_theta/checkpoints/flax_checkpoint.py (hunk 4 above)
Comment (blester125) [reply git_theta/checkpoints/flax_checkpoint.py:20]: fixed

File: tests/end2end/inprocess/test.sh (hunk 6 above)
Comment (blester125) [reply tests/end2end/inprocess/test.sh:21]: One reason is that verfiy.py isn't written in a way that makes importing it into other scripts easy, plus doing an import of a file in a different directory without having things setup as an installable package (__init__.py, setup.py, etc) can be weird.

Also the main runner for e2e tests is designed to run shell scripts so we would need to have a special case for this one and duplicate the color prints and things like that.

File: git_theta/filters.py (hunk 7 above)
Comment (blester125) [reply git_theta/filters.py:1]: No, they just got moved so they are more reusable.

File: tests/end2end/inprocess/test.py (hunk 8 above)
Comment (blester125) [reply tests/end2end/inprocess/test.py:25]: They do get materialized on disk, it's just in the "git future" so they do live on disk, but it's a question of if the git working directory has the right commit checked out. So the path would be materialized in the working tree if you checked out the next commit, but it isn't checked out right now.

I think the main question is just should the checkout step be an optional part of the save/load function (if we force it we would do a bunch of extra I/O we don't need to) or if the caller should just make a checkout afterwards.

File: git_theta/api.py (hunk 1 above)
Comment (craffel) [reply git_theta/api.py:18]: Yeah, I think I am philosophically opposed to making the default to have an uninformative commit message - we want git-theta to be more than torch.save on steroids.

File: tests/end2end/inprocess/test.py (hunk 8 above)
Comment (craffel) [reply tests/end2end/inprocess/test.py:25]: Should we have an arg (true by default seems to make sense to me) that specifies whether the checkout should be done after a save/load?

File: git_theta/api.py (hunk 1 above)
Comment (blester125) [reply git_theta/api.py]: I made it required

File: tests/end2end/inprocess/test.py (hunk 8 above)
Comment (blester125) [reply tests/end2end/inprocess/test.py:27]: I added flags that will keep the on-disk data sync'd to the in-memory model

File: git_theta/api.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+
+def save(
+    state_dict,
+    path: str,
+    commit_msg: str,
+    checkpoint_type: str = "pytorch",
+    materialize: bool = False,
+) -> git.Commit:
+    """Save a model using git-theta without writing it to disk."""
```
Comment (craffel) [inline git_theta/api.py:20]: Should we make these into proper docstrings if we want people to be calling them from Python?

File: git_theta/api.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    return sha
+
+
+def load(
+    sha_or_tag: Union[str, git.Commit],
+    path: str,
+    checkpoint_type: str = "pytorch",
+    checkout: bool = False,
+):
+    """Load a model from git-theta without having it checked out."""
```
Comment (craffel) [inline git_theta/api.py:59]: Probably change the docstring given the new checkout arg

File: tests/end2end/inprocess/test.py (hunk 8 above)
Comment (craffel) [reply tests/end2end/inprocess/test.py:27]: ... should they default to True?

File: git_theta/api.py (hunk 9 above)
Comment (blester125) [reply git_theta/api.py]: Updated

File: git_theta/api.py (hunk 10 above)
Comment (blester125) [reply git_theta/api.py]: Updated

File: tests/end2end/inprocess/test.py (hunk 8 above)
Comment (blester125) [reply tests/end2end/inprocess/test.py:31]: I don't think they should default to True, this isn't the way we would implement save/load if the main goal is keeping memory and disk in sync. If we wanted to do that our implementations would be:

```python
def save(...):
  ckpt.save(path)
  repo.git.commit(msg)
  
def load(...):
  repo.git.checkout(sha_or_tag)
  ckpt.load(path)
```

When you don't require the working tree has to be in sync with the in-memory then you only need 1 write to git storage for saving and 1 read from git storage for loading. When you force the checkout syncing you add an extra read/write of the model so you end up with the same cost as the simple implementations above (which are really basically wrappers around making/checking out commits, and seems less ergonomic than using git directly).

My thought was that this api would be used for things like, as you are training you make a commit/tag each time your model does better on the dev set and then at the end you load the best performing model and run it on the test set. Since this is a long running program but automatic program we don't care that the working tree is different. Then once the training is done, you can jump around in git history to get different model checkpoints instead of having multiple copies indexed by the step.

Comment (craffel) [reply tests/end2end/inprocess/test.py:31]: I see. I guess I just worry about someone running a .save and expecting that this implies that the checkpoint is actually saved to disk where they expect it to be (as would happen with a torch.save or whatever). Not sure what the right way to communicate that this is *not* the case.

Comment (blester125) [reply tests/end2end/inprocess/test.py:31]: I could add a log warning/info that is like 

```
You have saved a model to ${path} directly in git. You are no longer at HEAD, run `git checkout ${sha}` or `git checkout HEAD` to see the model in the working tree.
```

I can see how people might be confused, but just seems weird to have functions that are basically pointless unless non-default fault flags are set lol

Comment (craffel) [reply tests/end2end/inprocess/test.py:31]: Yeah, I understand the point that these functions are only beneficial/make sense with "False" default, but I think the "False" behavior is probably not what users are expecting from functions called "save"/"load". Is there any way to do the git-to-working-directory copy asynchronously (i.e. not blocking the Python process that did the save/load )or is that a bad idea? Otherwise, maybe it would be clearer if the functions were called "write_to_git"/"load_from_git" or something? Or maybe I am making wrong assumptions about user expectations, not sure.

Comment (blester125) [reply tests/end2end/inprocess/test.py:31]: Good idea, I renamed them to `save_to_git` and `load_from_git`

**Review by craffel — APPROVED** [review]
